harmonica/synthetic/surveys.py

```diff
@@ -0,0 +1,145 @@
+    """
+    Create a synthetic ground survey
+
+    The observation points are sampled from the South Africa gravity dataset.
+    Only a portion of the original survey is sampled and its region is rescaled to the
+    passed ``region``.
+
+    Parameters
+    ----------
+    region : tuple or list
+        Boundaries of the synthetic region where the observation points will be located
+        in the following order: (``east``, ``west``, ``south``, ``north``, ...). All
+        subsequent boundaries will be ignored. All boundaries should be in Cartesian
+        coordinates and in meters.
+    cut_region : tuple (optional)
+        Region to reduce the extension of the survey. Must be boundaries of the original
+        survey, in degrees.
+
+    Returns
+    -------
+    survey : :class:`pandas.DataFrame`
+        Dataframe containing the coordinates of the observation points and their
+        elevation Cartesian coordinates for the synthetic model. All coordinates and
+        altitude are in meters.
+
+    See also
+    --------
+    datasets.fetch_south_africa_gravity: Fetch gravity station data from South Africa.
+    """
+    # Sanity checks for region and cut_region
+    check_region(region[:4])
+    check_region(cut_region)
+    # Fetch ground gravity survey from South Africa
+    survey = fetch_south_africa_gravity()
+    # Cut the region into the cut_region, project it with a mercator projection to
+    # convert the coordinates into Cartesian and move this Cartesian region into the
+    # passed region
+    survey = _adecuate_survey(survey, region, cut_region)
+    return survey
+
+
+def _adecuate_survey(survey, region, cut_region):
+    """
+    Cut, project and move the original survey to the passed region
+
+    Parameters
+    ----------
+    survey : :class:`pandas.DataFrame`
+        Original survey as a :class:`pandas.DataFrame` containing the following columns:
+        ``longitude``, ``latitude`` and ``elevation``. The ``longitude`` and
+        ``latitude`` must be in degrees and the ``elevation`` in meters.
+    region : tuple or list
+        Boundaries of the synthetic region where the observation points will be located
+        in the following order: (``east``, ``west``, ``south``, ``north``, ...). All
+        subsequent boundaries will be ignored. All boundaries should be in Cartesian
+        coordinates and in meters.
+    cut_region : tuple (optional)
+        Region to reduce the extension of the survey. Must be boundaries of the original
+        survey, in degrees.
+
+    Returns
+    -------
+    survey : :class:`pandas.DataFrame`
+        Dataframe containing the coordinates of the observation points and their
+        elevation Cartesian coordinates for the synthetic model. All coordinates and
+        altitude are in meters.
+    """
+    # Cut the data into the cut_region
+    inside_points = inside((survey.longitude, survey.latitude), cut_region)
+    survey = survey[inside_points].copy()
```

Comment:
By the way, I'm struggling with a pandas warning if I don't copy the `survey` DataFrame.
If I remove the `.copy` I get the following warning:

```traceback
/home/santi/.miniconda3/envs/harmonica/lib/python3.7/site-packages/numba/types/containers.py:3
  /home/santi/.miniconda3/envs/harmonica/lib/python3.7/site-packages/numba/types/containers.py:3: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
    from collections import Iterable

harmonica/tests/test_synthetic_surveys.py::test_ground_survey
harmonica/tests/test_synthetic_surveys.py::test_airborne_survey
harmonica/tests/test_synthetic_surveys.py::test_cut_region_ground_survey
harmonica/tests/test_synthetic_surveys.py::test_cut_region_airborne_survey
  /home/santi/git/harmonica/harmonica/synthetic/surveys.py:130: SettingWithCopyWarning: 
  A value is trying to be set on a copy of a slice from a DataFrame.
  Try using .loc[row_indexer,col_indexer] = value instead
  
  See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
    survey.longitude.values, survey.latitude.values

harmonica/tests/test_synthetic_surveys.py::test_ground_survey
harmonica/tests/test_synthetic_surveys.py::test_airborne_survey
harmonica/tests/test_synthetic_surveys.py::test_cut_region_ground_survey
harmonica/tests/test_synthetic_surveys.py::test_cut_region_airborne_survey
  /home/santi/git/harmonica/harmonica/synthetic/surveys.py:139: SettingWithCopyWarning: 
  A value is trying to be set on a copy of a slice from a DataFrame.
  Try using .loc[row_indexer,col_indexer] = value instead
  
  See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
    ) + w

harmonica/tests/test_synthetic_surveys.py::test_ground_survey
harmonica/tests/test_synthetic_surveys.py::test_airborne_survey
harmonica/tests/test_synthetic_surveys.py::test_cut_region_ground_survey
harmonica/tests/test_synthetic_surveys.py::test_cut_region_airborne_survey
  /home/santi/git/harmonica/harmonica/synthetic/surveys.py:142: SettingWithCopyWarning: 
  A value is trying to be set on a copy of a slice from a DataFrame.
  Try using .loc[row_indexer,col_indexer] = value instead
  
  See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
    ) + s
```

Do you have any idea about what's going on here?

Reply:
You're right about keeping only the coordinates columns. In fact I ditch all other columns on the last line of `_adecuate_survey` (or `_cut_and_scale` as you propose).

Reply:
I moved the ditching of all the extra columns to each public function, before calling the `_cut_and_scale` function.

Reply:
:+1: